infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/KubernetesBrokerInitContainerApplier.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins;
+
+import com.google.common.annotations.Beta;
+import com.google.inject.Inject;
+import io.fabric8.kubernetes.api.model.Container;
+import io.fabric8.kubernetes.api.model.Pod;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironment;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalMachineConfig;
+import org.eclipse.che.api.workspace.server.wsplugins.model.PluginMeta;
+import org.eclipse.che.workspace.infrastructure.kubernetes.Names;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.wsplugins.brokerphases.BrokerEnvironmentFactory;
+
+/**
+ * Given a {@link InternalEnvironment} representing a workspace, adds the plugin broker as an init
+ * container to the workspace Pod. This is necessary if the workspace is created using emptyDir
+ * volumes (to allow the broker to add files to the workspace pod's volumes).
+ *
+ * <p>This API is in <b>Beta</b> and is subject to changes or removal.
+ *
+ * @author Angel Misevski
+ */
+@Beta
+public class KubernetesBrokerInitContainerApplier<E extends KubernetesEnvironment> {
+
+  private final BrokerEnvironmentFactory<E> brokerEnvironmentFactory;
+
+  @Inject
+  public KubernetesBrokerInitContainerApplier(
+      BrokerEnvironmentFactory<E> brokerEnvironmentFactory) {
+    this.brokerEnvironmentFactory = brokerEnvironmentFactory;
+  }
+
+  /**
+   * Apply plugin broker as init container to workspace environment. Workspace environment will have
+   * broker's configmap, machines, and volumes added in addition to the init container
+   *
+   * @param workspaceEnvironment
+   * @param runtimeID
+   * @param pluginsMeta
+   * @throws InfrastructureException
+   */
+  public void apply(
+      InternalEnvironment workspaceEnvironment,
```

Comment:
I think that we haven't used `workspaceEnvironment` word anywhere. Please consider changing it to `internalEnvironment`

Reply:
My intention in using `workspaceEnvironment` was to avoid confusion with `brokerEnvironment` and to support the intuition that we're basically copying environment info from the broker to the workspace. I do agree that immediately casting `workspaceEnvironment` to `kubernetesEnvironment` breaks that idea completely though. 

I've set the type of `workspaceEnvironment` to `E` and removed the cast. WDYT?

Reply:
> My intention in using workspaceEnvironment was to avoid confusion with brokerEnvironment

Makes sense